### ISSUE TITLE
Some debugger variable display issues

### ISSUE THREAD

**Issue by DeanRoddey** [rust-lang/rust#70542]
This may be a dup but I didn't see one right off hand.

Visual Studio Code: 1.43.2
Rust 1.42.0 
Toolchain: stable-x86_64-pc-windows-msvc
C/C++ VSC extension: 0.26.3
Rust RLS: 0.7.0

![VSCRustCharVal](https://user-images.githubusercontent.com/45377887/77855591-52cbc680-71bf-11ea-9226-774ab0d665f5.png)

**Not displaying char variables**

The debugger always shows an unspecified error for the value of char variables. Presumably this is some issue getting the right symbol info to the debugger?

**Optimizing even if in debug profile**

_[Here I think I need to retract this part of my complaint, since it appears that all of these I've seen so far are actually misleading messages, where the variable is reported as optimized out when really it's just not in scope.]_

Also, despite the optimization level being zero, locals keep getting optimized away, or at least the debugger thinks they are. I have seen others complaining about this as well, both on MSVC and Gdb so it doesn't seem specific to this tool chain. You can see in the image above a local claiming to have been optimized away, even though it's a debug build with optimization at zero.

Sometimes it's enough to make debugging fairly unproductive.

**Comment by DeanRoddey** [comment]
I would also throw in that the optimized variables thing is made a bit messier since it often shows that message even if it's just that the variable hasn't come into scope yet. Just not displaying variables that aren't in scope would probably be better, but the alternative would be to display a somewhat more accurate message like 'not in scope'.

**Comment by jonas-schievink** [comment]
We cannot control your debugger frontend, only how rustc emits debuginfo

**Comment by jonas-schievink** [comment]
The "variables display as optimized out" issue is https://github.com/rust-lang/rust/issues/46698.

The issue with chars looks like an issue with your debugger. Please check if that works with GDB or LLDB.

**Comment by DeanRoddey** [comment]
I understand you can't control the debugger, but all debuggers I've used don't display variables until/unless they are in scope. It seems a bit unlikely this one does otherwise. Presumably that is controlled by when the symbols are emitted, right?

**Comment by DeanRoddey** [comment]
Going back again, it seems to me now that all of the optimized out variable errors were really misleading and they were really just not in scope. So, for now, I'm going to delete that part of my complaint since I'm not sure if I really have any legit issues of that sort now. It's possible a lot of the complaints you've gotten may be of that sort.

**Comment by CryZe** [comment]
Alright, so this is happening for me and it can't be a recent regression as the same problem happens for Rust 1.20:
![https://i.imgur.com/IROTwON.png](https://i.imgur.com/IROTwON.png)

and additionally it also happens for C++ code (though differently):
![https://i.imgur.com/nYADJWK.png](https://i.imgur.com/nYADJWK.png)

I'm assuming it's just the debugger that changed how it behaves.

**Comment by CryZe** [comment]
What however seems weird is that the variable c suddenly appears out of nowhere while the others already exist:

![https://i.imgur.com/X16ykYB.gifv](https://i.imgur.com/X16ykYB.gif)

Also this time with Visual Studio directly instead of potentially flawed "third party extensions" (though the behavior is exactly the same in the extension).

**Comment by wesleywiser** [comment]
Visited during wg-debugging triage. I see three issues mentioned in this thread:

1. `char` values cannot be shows.
  - This was fixed by #89887
2. Variables sometimes reported as optimized out
  - There are various, more specific reports of this such as #46698
3. Weird issues with variable scoping
  - I've filed #112798 to track this. 

Closing due to the first issue being fixed and the other issues having more specific tracking issues.
